src/jxl.imageio/jxlinput.cpp

```diff
@@ -0,0 +1,358 @@
+{
+    return new JxlInput;
+}
+
+OIIO_EXPORT const char* jxl_input_extensions[] = { "jxl", nullptr };
+
+OIIO_PLUGIN_EXPORTS_END
+
+bool
+JxlInput::valid_file(Filesystem::IOProxy* ioproxy) const
+{
+    DBG std::cout << "JxlInput::valid_file()\n";
+
+    // Check magic number to assure this is a JPEG file
+    if (!ioproxy || ioproxy->mode() != Filesystem::IOProxy::Read)
+        return false;
+
+    uint8_t magic[128] {};
+    const size_t numRead = ioproxy->pread(magic, sizeof(magic), 0);
+    if (numRead != sizeof(magic))
+        return false;
+
+    JxlSignature signature = JxlSignatureCheck(magic, sizeof(magic));
+    switch (signature) {
+    case JXL_SIG_CODESTREAM:
+    case JXL_SIG_CONTAINER: break;
+    default: return false;
+    }
+
+    DBG std::cout << "JxlInput::valid_file() return true\n";
+    return true;
+}
+
+bool
+JxlInput::open(const std::string& name, ImageSpec& newspec,
+               const ImageSpec& config)
+{
+    DBG std::cout << "JxlInput::open(name, newspec, config)\n";
+
+    ioproxy_retrieve_from_config(config);
+    m_config.reset(new ImageSpec(config));  // save config spec
+    return open(name, newspec);
+}
+
+bool
+JxlInput::open(const std::string& name, ImageSpec& newspec)
+{
+    DBG std::cout << "JxlInput::open(name, newspec)\n";
+
+    m_filename = name;
+
+    DBG std::cout << "m_filename = " << m_filename << "\n";
+
+    if (!ioproxy_use_or_open(name)) {
+        DBG std::cout << "ioproxy_use_or_open returned false\n";
+        return false;
+    }
+
+    Filesystem::IOProxy* m_io = ioproxy();
+    std::string proxytype     = m_io->proxytype();
+    if (proxytype != "file" && proxytype != "memreader") {
+        errorfmt("JPEG XL reader can't handle proxy type {}", proxytype);
+        return false;
+    }
+
+    m_decoder = JxlDecoderMake(nullptr);
+    if (m_decoder == nullptr) {
+        DBG std::cout << "JxlDecoderMake failed\n";
+        return false;
+    }
+
+    m_runner = JxlResizableParallelRunnerMake(nullptr);
+    if (m_runner == nullptr) {
+        DBG std::cout << "JxlThreadParallelRunnerMake failed\n";
+        return false;
+    }
+
+    JxlDecoderStatus status = JxlDecoderSetParallelRunner(
+        m_decoder.get(), JxlResizableParallelRunner, m_runner.get());
+    if (status != JXL_DEC_SUCCESS) {
+        DBG std::cout << "JxlDecoderSetParallelRunner failed\n";
+        return false;
+    }
+
+    status
+        = JxlDecoderSubscribeEvents(m_decoder.get(),
+                                    JXL_DEC_BASIC_INFO | JXL_DEC_COLOR_ENCODING
+                                        | JXL_DEC_FRAME | JXL_DEC_FULL_IMAGE);
+    if (status != JXL_DEC_SUCCESS) {
+        DBG std::cout << "JxlDecoderSubscribeEvents failed\n";
+        return false;
+    }
+
+    std::vector<uint8_t> jxl;
+
+    DBG std::cout << "proxytype = " << proxytype << "\n";
+    if (proxytype == "file") {
+        size_t size = m_io->size();
+        DBG std::cout << "size = " << size << "\n";
+        jxl.resize(size);
```

Comment:
Turns out that std::vector.resize() will initialize the memory, which is a lot if this is a large image. I would recommend that instead you use:

    std::unique_ptr<uint8_t[]> jxl;
    ...
    jxl.reset(new uint8_t[size]);

If you only need that fixed size and aren't otherwise growing or shrinking the vector, it will all work just like the vector did, including auto-freeing when it leaves scope. But it won't needlessly touch the memory an extra time by zeroing it out.

Reply:
Sure, good point. Will try.